src/lib.rs

```diff
@@ -131,26 +131,70 @@ pub fn keccakf(a: &mut [u64; PLEN]) {
     }
 }
 
-fn setout(src: &[u8], dst: &mut [u8], len: usize) {
-    dst[..len].copy_from_slice(&src[..len]);
-}
+/// Total number of lanes.
+const PLEN: usize = 25;
+
+#[derive(Default, Clone)]
+struct Buffer([u64; PLEN]);
+
+impl Buffer {
+    fn inner(&mut self) -> &mut [u64; PLEN] {
+        &mut self.0
+    }
+
+    #[cfg(target_endian = "little")]
+    #[inline]
+    fn execute<F: FnOnce(&mut [u8])>(&mut self, offset: usize, len: usize, f: F) {
+        let buffer: &mut [u8; PLEN * 8] = unsafe { ::core::mem::transmute(&mut self.0) };
+        f(&mut buffer[offset..][..len]);
+    }
 
-fn xorin(dst: &mut [u8], src: &[u8]) {
-    assert!(dst.len() <= src.len());
-    let len = dst.len();
-    let mut dst_ptr = dst.as_mut_ptr();
-    let mut src_ptr = src.as_ptr();
-    for _ in 0..len {
-        unsafe {
-            *dst_ptr ^= *src_ptr;
-            src_ptr = src_ptr.offset(1);
-            dst_ptr = dst_ptr.offset(1);
+    #[cfg(target_endian = "big")]
+    #[inline]
+    fn execute<F: FnOnce(&mut [u8])>(&mut self, offset: usize, len: usize, f: F) {
```

Comment:
It seems a bit wasteful to actually convert to and back again, wouldn't it be better to modify `f`s in a endianess-agnostic way?

Reply:
I agree and that's what I've tried in a first place. But then I realised that the number of operations (multiplication and rem) to calculate the correct indexes of bytes is even more expensive and simply swapping the endianess of those we use